### DISCUSSION TITLE
Model 의 Choices 정의 방식 변경

### DISCUSSION THREAD

**Discussion by twins1040** [8percent/styleguide#37] [Ideas]
## Abstract
- `django-model-utils Choices` 에서 `Django Choices` 로 모듈을 변경하여 사용할 것을 제안합니다.
- 추가로 Choices 네이밍 룰을 제안합니다. 
  - 단수형 사용 (`LoanType` :o:, `LoanTypes` :x:)
  - Choices, Choice 는 끝에 붙이지 않음 (`Status` :o:, `StatusChoices` :x:)

## Rationale
- Django 3.0 으로 넘어오면서 3rd party 라이브러리인 `django-model-utils Choices`를 대체할 수 있는 Django Choices 클래스가 추가되었습니다. https://django-model-utils.readthedocs.io/en/latest/utilities.html#choices
- 기존 방식
  ```Python
  class Loan(models.Model):

      LOAN_TYPE = Choices(
          (1, 'RESIDUAL', '부동산 잔금대출'),
          (2, 'MORTGAGE', '부동산 담보대출'),
          (3, 'PROJECT_FINANCING', '프로젝트 파이낸싱'),
      )

      loan_type = models.PositiveSmallIntegerField(
          '대출종류',
          choices=LOAN_TYPE,
          default=LOAN_TYPE.MORTGAGE,
      )
  ```
- 신규 방식
  ```Python
  class Loan(models.Model):
  
      class LoanType(models.IntegerChoices):
          RESIDUAL = 1, '부동산 잔금대출'
          MORTGAGE = 2, '부동산 담보대출'
          PROJECT_FINANCING = 3, '프로젝트 파이낸싱'

      loan_type = models.PositiveSmallIntegerField(
          '대출종류',
          choices=LoanType.choices,
          default=LoanType.MORTGAGE.value,
      )    
  ```
- 변경 시 장점
  - 별도의 `from model_utils import Choices` 를 import 할 필요가 없습니다. (models 모듈에 포함)
  - 각각의 choice 에 대한 자동완성 및, 사용처 추적이 가능해 집니다.
    - 기존 Choices 는 python name 을 String 으로 정의했기 때문에 symbol 등록이 되지 않았습니다.
    - 따라서 IDE 가 해당 symbol 의 존재를 몰라서 추적 및 자동완성이 동작하지 않았고 매번 입력해주어야 했습니다.
- 변경 시 단점 
  - class 와 elemenet 값 사용 시 각각 `.choices` 와 `.value` 를 붙여주어야 합니다. (Enum 에 익숙하다면 오히려 좋을 수도?)
  - IntegerChoices, TextChoices 를 구분해서 사용해야 합니다.
- 네이밍 룰 관련
  - 현재 사용되고 있는 Choices 인스턴스의 변수명들
    1.`COMPANY_SIZE` 와 같이 Choice 관련 단어 없이 단어 그 자체가 Choice 를 의미
    2.`CATEGORIES`, `TYPES` 와 같이 Choice 관련 단어는 없지만, 복수형으로 작성
    3.`StatusChoices` 와 같이 Choices 붙임
  - 공식 문서 참조
    - 대부분 단수형을 사용하고 있고, Choices 를 붙이지 않는 것으로 보입니다.
    - Django 공식 문서 https://docs.djangoproject.com/en/3.2/ref/models/fields/#enumeration-types
      - `YearInSchool`: 신입생, 졸업생 등 구분
      - `Vehicle`: 차의 종류
      - `Suit`: 카드의 문양 목록
      - `MoonLandings`: 달 착륙 일 목록 (여기서만 복수형 사용?)
      - `Answer`: YES, NO
    - python Enum 공식 문서 (Choices 는 Enum 의 자식임) https://docs.python.org/3/library/enum.html#module-enum
      - `Color`: RED, GREEN, YELLOW 
      - `Shake`: VANILLA, MINT
      - `Shape`: SQUARE, DIAMOND
      - `Ordinal`: North, South, East, West
    - stack overflow .Net 관련 답변 https://stackoverflow.com/questions/1335959/singular-or-plural-for-enumerations
      - Do not use an Enum suffix on Enum type names.
      - Use a singular name for most Enum types, but use a plural name for Enum types that are bit fields.
    - Java 관련 Enum 답변 https://stackoverflow.com/questions/15755955/naming-of-enums-in-java-singular-or-plural
      - singular 사용 
  - Enum 을 단수로 선언해야 하는 이유 심층 분석
    - https://docs.python.org/3/library/enum.html#enum-members-aka-instances 를 참조해 추측해볼 수 있습니다.
    - LoanType.MORTGAGE 의 형태를 잠시 잊고 만약 아래와 같이 MORTGAGE 가 LoanType 의 인스턴스라면 어떨까요?
      - MORTGAGE = LoanType() or MORTGAGE = LoanTypes()
    - 전자가 자연스럽습니다. 사람을 각각 정의할 때 Persons('홍길동') 으로 정의하지 않고 Person('홍길동') 으로 정의하는 것과 마찬가지 입니다.
    - 그런데 실제로 MORTGAGE 는 LoanType 의 인스턴스입니다.
    - Python 의 Enum 타입은 EnumMeta 클래스를 통해 각각의 멤버들을 Enum 클래스로 인스턴스화 한 뒤에 __new__ 메서드를 치환해버리는 식으로 싱글톤 패턴을 구현하고 있기 때문입니다.
    - 위와 같은 이유로 Enum 클래스에 name, value 가 정의 된 것을 볼 수 있습니다.
    - 또한 사용하면 안되는 방식이지만 LoanType.RESIDUAL.MORTGAGE 와 같은 접근도 가능합니다.
    - 결론적으로 Enum 은 instance 인 멤버들의 class 그 자체이기 때문에 단수로 명명하는 것이 맞다고 보여집니다.
    - LoanType.choices 가 실질적으로 리스트형을 표현한다고 볼 수 있습니다.
 
## Specification
```Python
class Loan(models.Model):

    class LoanType(models.IntegerChoices):
        RESIDUAL = 1, '부동산 잔금대출'
        MORTGAGE = 2, '부동산 담보대출'
        PROJECT_FINANCING = 3, '프로젝트 파이낸싱'

    loan_type = models.PositiveSmallIntegerField(
        '대출종류',
        choices=LoanType.choices,
        default=LoanType.MORTGAGE.value,
    )    
```
- Model 내 Choices 선언 시, models.Choices 를 상속받은 클래스를 사용합니다.
- 클래스 명은 단수형을 사용합니다.

자세한 내용은 https://docs.djangoproject.com/en/3.2/ref/models/fields/#enumeration-types 참조

**Comment by donghyun0218** [comment]
외부 패키지의 의존성도 줄일 수 있다는 장점이 있군요. 좋은 것 같습니다.

한 가지 궁금한 점은 현재 저희 코드 베이스 내에서는 `django-model-utils Choices` 를 모델을 위한 용도가 아닌, 상수로써도 사용을 하는 경우가 있습니다.

이러한 경우에 대해서도 `django-model-utils Choices` 가 아닌 `Django Choices` 를 사용하는 것을 제안하시나요??

**Reply by twins1040** [reply]
네 저는 이 경우도 함께 바꾸면 좋을 것 같습니다.

**Reply by cbscsm** [reply]
기존 코드 상수 중에는 `eight.constants.ChoiceConstants`(`candv.Constants`의 서브클래스)를 쓰는 경우도 많이 있는 것 같습니다. 상수목적 choice에 대해서도 한번 정리해보면 좋을 것 같네요!

**Comment by Kirade** [comment]
네이밍, 사용방식에 대해 이견있는 부분 없습니다.

변경하는것에 동의합니다.

**Comment by kyungjunleeme** [comment]
네, 변경하는것에 동의합니다. 
- 언급해주신 IntegerChoices, TextChoices 를 구분해서 사용해야 합니다. 

이 부분만 유의하면 될 것 같습니다. 정리해주셔서 감사합니다

**Comment by wer153** [comment]
이견없이 동의합니다.

**Comment by cbscsm** [comment]
동의합니다.
